### PR TITLE
SmallPermutationGroups

### PR DESCRIPTION
**Pull request by dwbump**
### 📚 Description

GAP has a database of Small Groups identified by a pair of numbers [n,k] where n is the order and k is an integer used to differentiate different groups of the same order. Thus k is bounded by oeis(1)[n], the number of isomorphism classes of groups of order n. Since Sage PermutationGroups have many useful methods, it is good to have PermutationGroup realizations of GAP SmallGroups. The docstring verifies that the new SmallPermutationGroup class is able to construct all groups of order <= 64, though of course groups of order >64 are also available.

### 📝 Checklist

- [ x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [ ] I have linked an issue or discussion.
- [ x] I have created tests covering the changes.
- [ x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by dimpase — COMMENTED** [review]
Can you use libgap rather than gap? I.e. 
```
self._gap_small_group = libgap.SmallGroup(order,gap_id)
```

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 1]
```diff
@@ -87,7 +89,9 @@
 from pathlib import Path
 
 from sage.rings.all import Integer
+from sage.interfaces.gap import gap
```
Comment (dimpase) [inline src/sage/groups/perm_gps/permgroup_named.py]: please no more dependencies on pexpect gap. We are trying to get rid of it in #26902

**Review by dimpase — CHANGES_REQUESTED** [review]

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 1 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: If we do the suggested self._gap_small_group = libgap.SmallGroup(order,gap_id) then we lose some functionality, needed for an application (FusionDouble, probably an upcoming PR). We need the methods conjugacy_classes_representatives and centralizer.

I haven't figured out why these do not work when we use libgap. I tried making SmallPermutationGroup inherit from PermutationGroup_generic since that class has the needed methods.

Comment (dimpase) [reply src/sage/groups/perm_gps/permgroup_named.py]: is `FusionDouble` a new functionality, not in Sage yet?

Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: Yes, analogous to FusionRing one can construct the Fusion Rings of the Drinfeld Double of a finite group.

**Comment by dimpase** [comment]
is this code available somewhere? I can try to port it to libgap.

**Comment by dwbump** [comment]
> is this code available somewhere? I can try to port it to libgap.

Working code exists (written by Wenqi Li and myself) but work needs to be done before it is ready for review. Particularly, I think Willie Aboumrad and Travis Scrimshaw were going to try to refactor it into the existing FusionRing class.

I think what is needed now is an understanding as to why changing gap to libgap here breaks the centralizer and conjugacy_class_representatives methods.

**Comment by dimpase** [comment]
On Thu, 2 Mar 2023, 19:32 Daniel Bump, ***@***.***> wrote:

> is this code available somewhere? I can try to port it to libgap.
>
> Working code exists (written by Wenqi Li and myself) but work needs to be
> done before it is ready for review. Particularly, I think Willie Aboumrad
> and Travis Scrimshaw were going to try to refactor it into the existing
> FusionRing class.
>
> I think what is needed now is an understanding as to why changing gap to
> libgap here breaks the centralizer and conjugacy_class_representatives
> methods.
>

it could be that the latter need porting to libgap.
Can you point at more precise code location?

**Comment by dwbump** [comment]
it could be that the latter need porting to libgap.
Can you point at more precise code location?

**Comment by dwbump** [comment]
> it could be that the latter need porting to libgap.
> Can you point at more precise code location?

I added a doctest at line 3479 of the file, to compute some centralizers. So if you uncomment line 3487, changing gap to libgap, that breaks the doctest. The centralizer and conjugacy_classes_representatives methods are implemented in the PermutationGroup_generic class, in permgroup.py.

**Comment by dimpase** [comment]
thanks - I'll look at it after the conference in Luminy I'm attending ends, tomorrow.

**Comment by ChrisJefferson** [comment]
Some issues with this PR:

It might be worth mentioning in the docs that this group isn't fixed -- while it will always be isomorphic to the SmallGroup in question, it can change over time. Also, in many cases [n,k] will be a permutation group on n points -- but that won't always be true.

Also, there is a practical issue. `Image(IsomorphismPermGroup(g))` doesn't return what you want (always) -- as it can return a larger group (this behaviour may be changed in a future version of GAP to what you want, but it isn't that right now -- https://github.com/gap-system/gap/issues/5355 ).

Instead you want to call `Image(IsomorphismPermGroup(g),g)`

**Comment by dimpase** [comment]
> Instead you want to call `Image(IsomorphismPermGroup(g),g)`

fixed by a57a9c0ca0d38669491f284fe15e0d37b82cfddd

**Comment by dwbump** [comment]
> It might be worth mentioning in the docs that this group isn't fixed -- while it will always be isomorphic to the SmallGroup in question, it can change over time. Also, in many cases [n,k] will be a permutation group on n points -- but that won't always be true.

I edited the docstring to mention these points. The other comment on Image was addressed by a commit by dimpase.

**Comment by dimpase** [comment]
the immediate `libgap` issue is that the group in the docstring is not converted to a permutation group, it's a just a GAP integer `6`.  Is it related to `in many cases [n,k] will be a permutation group on n points -- but that won't always be true.` ?

**Comment by dwbump** [comment]
> the immediate libgap issue is that the group in the docstring is not converted to a permutation group, it's a just a GAP integer 6. Is it related to in many cases [n,k] will be a permutation group on n points -- but that won't always be true. ?

It doesn't seem related to this point. I say this because for some k the group [12,k] is of degree 12 and for other k it is of degree 7, but in every case the libgap version of the group fails this test.

**Comment by dimpase** [comment]
On Fri, Mar 3, 2023 at 5:20 PM Daniel Bump ***@***.***> wrote:

> the immediate libgap issue is that the group in the docstring is not
> converted to a permutation group, it's a just a GAP integer 6. Is it
> related to in many cases [n,k] will be a permutation group on n points --
> but that won't always be true. ?
>
> It doesn't seem related to this point. I say this because for some k the
> group [12,k] is of degree 12 and for other k it is of degree 7, but in
> every case the libgap version of the group fails this test.
>

I don't understand how the gap version works.

Ineed, you pass the number of moved points to the constructor of
PermutationGroup_generic, but its signature is

def __init__(self, gens=None, gap_group=None, canonicalize=True,
domain=None, category=None):

and in the code I don't see whether there is any provision for handling
this number of moved points.

Could you please explain what you're trying to do here. As well, we better
pass gap_group, not gens.

And it would help if you used explicit parameter names...

**Comment by dwbump** [comment]
> Ineed, you pass the number of moved points to the constructor of PermutationGroup_generic, but its signature is
> def __init__(self, gens=None, gap_group=None, canonicalize=True, domain=None, category=None):

You are right, for the libgap group we do not want to pass the degree of the permutation representation, we want to pass the gap group. This seems to be the key to fixing the libgap group, which I'll have to take care of later but thanks for noticing this.

**Comment by dimpase** [comment]
Can't post a patch, using libgap, now, but can do it in few hours.
New doctests will need a trivial adjustment, too.

**Comment by dwbump** [comment]
> Can't post a patch, using libgap, now, but can do it in few hours.

I can do this (though not at the moment).

> New doctests will need a trivial adjustment, too.

I can do this also.

**Comment by dwbump** [comment]
I pushed changes in response to Dima's comment that we were passing the wrong thing to the libgap group __init__ method. I still want to do more testing.

**Review by dimpase — APPROVED** [review]

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 2]
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
         ret = [self._m * i for i in reversed(range(self._n-1))]
         ret.append((self._n-1)*self._m - self._n)
         return tuple(sorted(ret, reverse=True))
+
+class SmallPermutationGroup(PermutationGroup_generic):
+    def __init__(self, order, gap_id):
+        r"""
```
Comment (tscrim) [inline src/sage/groups/perm_gps/permgroup_named.py]: I would move this doc to the class level and have a `TestSuite` doctest in `__init__`.

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 3]
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
...
+            False
+            sage: [H.centralizer(g) for g in H.conjugacy_classes_representatives()]
+            [Subgroup generated by [(1,2)(3,6)(4,5), (1,3,5)(2,4,6)] of (Group of order 6 and GAP Id 1 as a permutation group),
+            Subgroup generated by [(1,2)(3,6)(4,5)] of (Group of order 6 and GAP Id 1 as a permutation group),
+            Subgroup generated by [(1,3,5)(2,4,6), (1,5,3)(2,6,4)] of (Group of order 6 and GAP Id 1 as a permutation group)]
+        """
+        self._order = order
+        self._gap_id = gap_id
+        self._gap_small_group = libgap.SmallGroup(order,gap_id)
+        self._gap_permutation_group = self._gap_small_group.IsomorphismPermGroup().Image(self._gap_small_group)
```
Comment (tscrim) [inline src/sage/groups/perm_gps/permgroup_named.py]: Why do we need this attribute? It already exists in the base class (under a different name).

File: src/sage/groups/perm_gps/permgroup.py [hunk 4]
```diff
@@ -247,6 +247,8 @@ def from_gap_list(G, src):
     """
     # src is a list of strings, each of which is a permutation of
     # integers in cycle notation. It may contain \n and spaces.
+    if G is None:
+        return [eval(("[%s]"%g).replace(" ","").replace("\n","").replace(")(","),(")) for g in src]
```
Comment (tscrim) [inline src/sage/groups/perm_gps/permgroup.py]: This seems to no longer be used. Can we revert this change?

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 5]
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
...
+            Group of order 12 and GAP Id 4 as a permutation group
+        """
+        return "Group of order %s and GAP Id %s as a permutation group"%(self._order, self._gap_id)
+
+    def gap_small_group(self):
+        r"""
+        Gap realizes some small groups as PermutationGroup, others as PcGroups
+        (polycyclic groups). The :class:`SmallPermutationGroup` class always
+        returns a PermutationGroup, but in the process of creating this group
+        a gap SmallGroup is generated. This method returns that group.
```
Comment (tscrim) [inline src/sage/groups/perm_gps/permgroup_named.py]:
```suggestion
        Return the GAP small group object corresponding to ``self``.
        
        Gap realizes some small groups as PermutationGroup, others as PcGroups
        (polycyclic groups). The :class:`SmallPermutationGroup` class always
        returns a PermutationGroup, but in the process of creating this group
        a gap SmallGroup is generated. This method returns that group.
```

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 5 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: I made most of the changes requested by Travis, except one comment above that I am asking for clarification on.

Unfortunately if I add a testsuite to __init__ (currently commented out) there are many failures.

**Comment by tscrim** [comment]
That’s likely because there are some features missing that are expected. I will take a look tonight.

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 5 above)
Comment (tscrim) [reply src/sage/groups/perm_gps/permgroup_named.py]: I meant to put my comment below here.

You also have underindented the `EXAMPES::` block I think.

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 2 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: > That’s likely because there are some features missing that are expected. I will take a look tonight.

Thanks!

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 5 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: > You also have underindented the EXAMPES:: block I think.

Fixed.

Comment (tscrim) [reply src/sage/groups/perm_gps/permgroup_named.py]: The problem comes from the fact the permutation group expects that `_order` is a method. There are 2 ways to fix this. The first would be to change `_order` to, e.g., `_group_order` and override the `_order()` method to simply return `_group_order`. The second would be to override `order()` to simply return `self._order`. The method `_order()` only seems to be used in the `order()` method. So I would go with option 2.

Comment (tscrim) [reply src/sage/groups/perm_gps/permgroup_named.py]: You will also need to reset the `cardinality = order` alias.

Comment (tscrim) [reply src/sage/groups/perm_gps/permgroup_named.py]: So I added this at the end of the class and the `TestSuite` passed:
```python

    def order(self):
        return self._order

    cardinality = order
```

**Comment by dwbump** [comment]
It turned out that the TestSuite was failing because there was an attribute called `_order`. The current version passes the tests for me, but we still have not addressed one of Travis' comments:

>Why do we need this attribute? It already exists in the base class (under a different name).

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 5 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: > So I added this at the end of the class and the `TestSuite` passed:

I fixed this alternatively by just renaming the `_order` attribute `_n` and the testsuite passes.

>
> ```python
>     def order(self):
>         return self._order
> 
>     cardinality = order
> ```

**Comment by tscrim** [comment]
The one slight downside to doing it be renaming is that it now actually has to compute the order rather than take advantage of the fact is already "knows" it. However, since these are small groups, not such a big issue.

For my other comment, I am (implicitly) suggesting to change `self._gap_permutation_group` into a local variable.

**Comment by dwbump** [comment]
> For my other comment, I am (implicitly) suggesting to change self._gap_permutation_group into a local variable.

OK, I did it.

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 6]
```diff
@@ -88,6 +90,7 @@
 
 from sage.rings.all import Integer
 from sage.libs.gap.libgap import libgap
+from sage.groups.perm_gps.permgroup import from_gap_list
```
Comment (dimpase) [inline src/sage/groups/perm_gps/permgroup_named.py]: This import is not needed any more.

**Comment by dimpase** [comment]
> The one slight downside to doing it be renaming is that it now actually has to compute the order 
I think the groups from GAP's SmallGroups library already "know" their orders, i.e.
calling (lib)GAP's `Order()` on them is cheap.

If Sage still needs to compute it, it's bad design.

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 7]
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
         ret = [self._m * i for i in reversed(range(self._n-1))]
         ret.append((self._n-1)*self._m - self._n)
         return tuple(sorted(ret, reverse=True))
+
+class SmallPermutationGroup(PermutationGroup_generic):
+    r"""
+    A GAP SmallGroup, returned as a permutation group.
+
+    GAP contains a database of groups, each identified by its GAP id.
```
Comment (fingolfin) [inline src/sage/groups/perm_gps/permgroup_named.py]: GAP contains many databases of groups, with an id. This should clarify which one you are referring to: the small groups library? transitive groups? primitive groups? perfect groups? ...

if it is the small groups library (SGL), the correct name would be more like "small groups id" -- note that Magma uses the exact same ids

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 8]
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
...
+        PermutationGroup_generic.__init__(self, gap_group=gap_permutation_group)
+
+    def _repr_(self):
+        r"""
+        EXAMPLES::
+
+            sage: G = SmallPermutationGroup(12,4); G
+            Group of order 12 and GAP Id 4 as a permutation group
+        """
+        return "Group of order %s and GAP Id %s as a permutation group"%(self._n, self._gap_id)
```
Comment (fingolfin) [inline src/sage/groups/perm_gps/permgroup_named.py:3512]: Why "Id" and not "id" ?

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 9]
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
...
+
+    def gap_small_group(self):
+        r"""
+        Return the GAP small group object corresponding to ``self``.
+
+        Gap realizes some small groups as PermutationGroup, others as PcGroups
```
Comment (fingolfin) [inline src/sage/groups/perm_gps/permgroup_named.py]:
```suggestion
        GAP realizes some small groups as PermutationGroup, others as PcGroups
```
But actually, these groups are provided by the Small Groups Library, a package for GAP, not by the GAP core itself. Just mentioning in case you are for precise attribution...

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 10]
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
...
+        (polycyclic groups). The :class:`SmallPermutationGroup` class always
+        returns a PermutationGroup, but in the process of creating this group
+        a gap SmallGroup is generated. This method returns that group.
```
Comment (fingolfin) [inline src/sage/groups/perm_gps/permgroup_named.py]:
```suggestion
        a GAP SmallGroup is generated. This method returns that group.
```

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 9 above)
Comment (fingolfin) [reply src/sage/groups/perm_gps/permgroup_named.py]: You could also make it more precise: all solvable groups are PcGroups, the non-solvable ones permutation groups

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 8 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py:3512]: I revised the docstrings following these comments and pushed. Thanks!

**Comment by tscrim** [comment]
@dimpase @fingolfin Is this a positive review or not? It seems like the PR is approved, but the status badge wasn't changed. Also, what about https://github.com/sagemath/sage/pull/35225#issuecomment-1455059845?

**Comment by dimpase** [comment]
I'd like to hear about your opinion on https://github.com/sagemath/sage/pull/35225#issuecomment-1455059845 - why do you recompute known group orders?

**Comment by dimpase** [comment]
that's why I kept the tag "needs review" on.

**Comment by tscrim** [comment]
> I'd like to hear about your opinion on [#35225 (comment)](https://github.com/sagemath/sage/pull/35225#issuecomment-1455059845) - why do you recompute known group orders?

Which "you"? Me or @dwbump?

**Comment by tscrim** [comment]
So the current implementation uses what is available by default in Sage for permutation groups. In particular, it does eventually call GAP's `Size()` whenever the group is not generated by transpositions, but it checks this.

Now the order is passed to the constructor within Sage, but @dwbump has done is just store that as an attribute, which I said was okay as writing a custom `order()` method to just return said attribute introduces some added complexity of the code. In particular, it does the "computation" I just mentioned, which essentially leaves it up to GAP. Although I guess the underlying GAP group is actually the isomorphic permutation group, so perhaps GAP is having to do a bit more computation, but the groups are "small" so this should be quick.

Well...I would push the change overriding the `order()` method to my own branch based on @dwbump's branch, but I really don't want to do a whole new PR just for this. It seems less bad to do a PR based on this one, albeit still somewhat fragmented and should be merged at the same time...

**Comment by dwbump** [comment]
I added an order method that just returns `self._n`.

> So the current implementation uses what is available by default in Sage for permutation groups. In particular, it does eventually call GAP's `Size()` whenever the group is not generated by transpositions, but it checks this.

**Comment by vbraun** [comment]
Tries to download stuff from the internet:
```
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 1289, in first_terms
        return to_tuple(" ".join(flatten([self._field(a) for a in fields])))[:number]
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 1289, in <listcomp>
        return to_tuple(" ".join(flatten([self._field(a) for a in fields])))[:number]
                                          ^^^^^^^^^^^^^^
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 752, in _field
        for line in self.raw_entry().splitlines():
                    ^^^^^^^^^^^^^^^^
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 860, in raw_entry
        self.online_update()
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 730, in online_update
        self._raw = _fetch(url).split('\n\n')[2]
                    ^^^^^^^^^^^
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/databases/oeis.py", line 205, in _fetch
        raise IOError("%s\nerror fetching %s" % (msg, url))
    OSError: <urlopen error [Errno 110] Connection timed out>
    error fetching https://oeis.org/search?q=A000001&n=1&fmt=text
**********************************************************************
1 item had failures:
   1 of   8 in sage.groups.perm_gps.permgroup_named.SmallPermutationGroup
    [539 tests, 1 failure, 133.62 s]
----------------------------------------------------------------------
sage -t --long --random-seed=0 src/sage/groups/perm_gps/permgroup_named.py  # 1 doctest failed
----------------------------------------------------------------------
```

File: src/sage/groups/perm_gps/permgroup_named.py [hunk 11]
```diff
@@ -3435,3 +3437,104 @@ def codegrees(self):
...
+        (1,3)(2,5)(4,7)(6,9)(8,11)(10,12),
+        (1,4,8)(2,6,10)(3,7,11)(5,9,12))
+        sage: G.character_table()
+        [ 1  1  1  1  1  1]
+        [ 1 -1 -1  1  1 -1]
+        [ 1 -1  1  1 -1  1]
+        [ 1  1 -1  1 -1 -1]
+        [ 2  0 -2 -1  0  1]
+        [ 2  0  2 -1  0 -1]
+        sage: all(SmallPermutationGroup(n,k).id()==[n,k] for n in [1..64] for k in [1..oeis(1)[n]])
```
Comment (tscrim) [inline src/sage/groups/perm_gps/permgroup_named.py]: This test needs to be marked with `# optional - internet` and/or replaced by something that is manually inputted. (I would at least include part of the latter since it means it will get more regularly tested.)

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 11 above)
Comment (fingolfin) [reply src/sage/groups/perm_gps/permgroup_named.py]: Why does it query data from OEIS that is part of the very small groups library that is  eing wrapped here in the first place?

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 8 above)
Comment (fingolfin) [reply src/sage/groups/perm_gps/permgroup_named.py:3512]: It still says `Id` instead of `id` in several places

File: src/sage/groups/perm_gps/permgroup_named.py (hunk 11 above)
Comment (dwbump) [reply src/sage/groups/perm_gps/permgroup_named.py]: I replaced the oeis(1) by a call to libgap.NumberSmallGroups.

**Comment by dwbump** [comment]
Volker removed the Positive Review flair because of the oeis(1) call. I fixed that, so I added Needs Review. I am not sure why the pdf documentation is failing to build.

**Comment by tscrim** [comment]
LGTM modulo one trivial whitespace thing:
```diff
-sage: def numgps(n) : return ZZ(libgap.NumberSmallGroups(n))
+sage: def numgps(n : return ZZ(libgap.NumberSmallGroups(n))
```

I’m not sure why the pdf docs build is failing either. Did you check locally?

**Comment by dwbump** [comment]
> LGTM modulo one trivial whitespace thing:
> 
> ```diff
> -sage: def numgps(n) : return ZZ(libgap.NumberSmallGroups(n))
> +sage: def numgps(n : return ZZ(libgap.NumberSmallGroups(n))
> ```

I think you meant to delete the space before the colon and instead deleted the paren.

> I’m not sure why the pdf docs build is failing either. Did you check locally?

On my machine, make doc-pdf runs to completion and does not fail.

**Review by tscrim — APPROVED** [review]
Yes, that's right. Thank you.
